[SDK] - Proposal for samples generation via GO and in the SDK repo

camilamacedo86 opened the pull request:
NOTE: It has images and code example, so I'd suggest doing the review by seen the preview

estroz commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,292 @@
+
+## Proposal
+
+### I am as maintainer/contributor would like to:
+
+- be able to troubleshoot the code used to create the samples and easy improve them as solve issues that might be faced
+- no longer spend effort to update the samples on the release process
+- easily able to create samples based on the tutorial steps
+- ensure that the changes performed against the master will not introduce bugs in the steps required to build the samples and following up the tutorials
+- able to implement the code to generate the samples using the same approach used in the e2e tests (it is very helpful to keep the project maintained)
+- able to easily review pr's which has changes for the scaffolds (without the need to clone them locally)
+      
+### I am as operator developer would like to:
+- check the examples of projects and features provided to help me build my operators as know how they work
+- check the project which is the result of the tests performed in the tutorials provided
```
These are user stories, please organize them as such with the correct heading.

camilamacedo86 replied:
Hi @estroz, 

I did not get your comment. A user story technically is : 

> As a < type of user >, I want < some goal > so that < some reason >. 

So, it is organized. However, instead of duplicated many times `As a < type of user >, I want ` it is the title.

estroz replied:
Check out the [this example EP](https://github.com/operator-framework/enhancements/blob/master/enhancements/template.md#user-stories-optional) for correct formatting.

camilamacedo86 replied:
Hi @estroz, 

Sorry, but about this one, I feel that I cannot agree with. I will try to share my reasons and motivations. 

`User Story` is well defined in computer science and I do not agree that we should not follow its definition and common usage. Actually, I already shared in other moments and comments that IMO we need to write more "user stories" (e.g epic/rfe) and as it principle describes. 

Particularly, I really think that they are very helpful for I understand what is required to do and achieve. I mean `<who>` would able to do `<what>` with the implementation. And then, when I do not see it done as anywhere and in a very verbose way, for me, makes the process to understand the need harder than easier which also makes I believe that would be a problem for others as well. Note that if you looking for `user stories in computer science` you will find that it mainly as follows in everywhere. 

> As a `< type of user >, I want < some goal > so that < some reason >` 

See https://en.wikipedia.org/wiki/User_story. 

Regards `Check out the this example EP for correct formatting.`

See that the template describe:

> Detail the things that people will be able to do if this is implemented. Include as much detail as possible so that people can understand the "how" of the system. The goal here is to make this feel real for users without getting bogged down.

Which is exactly what was done here.

camilamacedo86 replied:
However, I agree that has to improvements. I will be working on that 👍

camilamacedo86 replied:
I am closing this one. It was improved. Thank you for your input.

estroz commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,292 @@
+
+### Risks and Mitigations
+
+N/A
```
There are no risks to generating complex example operators? What about copying parts of the reconcile loop into the generated operator and the copied code isn't updated?

camilamacedo86 replied:
The build of the samples is tested by the CI. However, is a NO GO run tests over the samples to ensure that they are runnable. Regards the inject code is wrong then, it is the risk of any code/change implementation against the project but OK, we can add that.

estroz reviewed (changes requested):
A few problems with this PR:
1. It mostly follows the EP template, but the `Proposal` section does not in a detrimental manner. Where are the implementation notes/details?
1. The EP is very light on technical detail. How are code changes to be made, ex. we update part of the reconcile loop? What is a `SampleContext` (there is a good example of one, but no discussion about its semantics in English).
1. There is no discussion as to how this will be tied into the SDK's release process, which is tangential to the generator's implementation but still relevant to this EP.

camilamacedo86 commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,292 @@
+
+This proposal is over the samples getting started to be generated on the SDK repo via go to bring more flexibilities and facilities to keep them maintained and improve their implementation and customizations. 
+
+By using Golang, we can easily inject code implementation as have helpers which brings a lot of facilities to customize them. Also, it allows a more readable implementation.  
```
Hi @estroz, 

See that has a PR with the implementation proposed here done already to cover all details. However, I will try to be more specific here.

camilamacedo86 commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,292 @@
+
+- Centralize the samples in the [operator-sdk-samples](https://github.com/operator-framework/operator-sdk-samples) repository generated on SDK repo
+- For each PR a validation should be made to ensure that all samples are update
+- Provide the SampleContext as part of the generation of the sample on SDK repo at the same way that provides the [TestContext](https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/utils/test_context.go#L36) 
```
It is true 💯

camilamacedo86 commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,292 @@
+
+## Summary
+
+This proposal is over the samples getting started to be generated on the SDK repo via go to bring more flexibilities and facilities to keep them maintained and improve their implementation and customizations. 
```
Accepted all with a few changes. See that the idea is not to use scripts but instead use code implementation done with Go.

camilamacedo86 commented:
Hi @estroz, 

Really thank you for your time and input. I addressed your suggestions and provide further details. I will re-check it more after to shape. However, feel free to see as well.

camilamacedo86 commented:
Hi @estroz, 

Please, feel free to re-check it.

varshaprasad96 commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+
+### Implementation Details/Notes/Constraints 
+
+- The samples should be build inside of `./testdata/<type>/` path
```
Just a quick doubt. Im not sure how KB does this. But, if we have different plugins for same type of operator, would we scaffold the project with default plugin version which is set.

camilamacedo86 replied:
At the moment we have only one version of the plugin. To create the samples we use the SDK binary and run the commands at the same way that users will do that. So, in the future, if we wish we can create samples for diff plugins versions since it means `operator-sdk --plugins="plugin version"`

jmrodri replied:
@varshaprasad96 @camilamacedo86 I would NOT scaffold out samples with multiple plugins. I would create the samples for our default use case only. And possibly for any complex use cases like we are talking about with webhooks. Honestly I can see 2 samples: 1) default easy peasy project. 2) one with all the options enabled to show the complex features.

camilamacedo86 replied:
@jmrodri it is not in the scope of the EP anyway. However, the stack allows us to do that if we wish in the future. I mean, has no limitation for that if we wish to do that. But I agree 100% with you.

varshaprasad96 commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+
+#### Generate samples via Go 
+
+- I am a maintainer/contributor, I want to be able to troubleshoot the code used to create the samples, so that I can easily fix the issues. 
```
If we are going to scaffold the operator from scratch and test all the functionalities, then would we be removing the existing e2e tests for go/ansible/helm? We would be doing the same steps twice.

asmacdo replied:
Using the samples for the e2e tests is an idea on the table, but when we talked about this initially we thought that the e2e tests should be out of scope for this proposal to keep it from becoming a huge task.

asmacdo commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+
+## Summary
+
+The [operator-sdk-samples](https://github.com/operator-framework/operator-sdk-samples) repo contains an example Go operator project for users to reference when building their own projects. These example becomes outdated quickly because it requires its own release process and manual updates.
```
```suggestion
The [operator-sdk-samples](https://github.com/operator-framework/operator-sdk-samples) repo contains an example Go operator project for users to reference when building their own projects. These examples become outdated quickly because it requires its own release process and manual updates.
```

asmacdo commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+
+In this way, it is no longer required keep the repository [operator-sdk-samples](https://github.com/operator-framework/operator-sdk-samples) and it should be deprecated. 
+
+**NOTE** The test utils as what is proposed here can fit well in upstream. It is not a Goal push these implementations to upstream, however, it definitely could be an ultimate end goal to keep in mind. 
```
```suggestion
**NOTE** The test utils as what is proposed here can fit well in upstream Kubebuilder. It is not a Goal push these implementations to Kubebuilder, however, it definitely could be an ultimate end goal to keep in mind. 
```

asmacdo commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+
+- Link the samples in the docs or aligned then with the tutorials
+- Provide advanced helpers or Boilerplates to mock the data that could be used in the tests and/or in the samples generation
+- Push this implementation to upstream
```
```suggestion
- Push this implementation to upstream Kubebuilder
```

asmacdo commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+#### Checking if samples are outdated via CI
+
+- I am a maintainer/contributor, I want to be able to check the samples update with the changes made, so that I do not need to spend effort on the samples after the release process
+- I am a maintainer/contributor, I want to able to check the samples updated in the diff of te github on the review process, so that I do not need to clone them locally
```
```suggestion
- I am a maintainer/contributor, I want to able to check the samples updated in the diff of the github on the review process, so that I do not need to clone them locally
```

asmacdo commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+
+#### Building samples on SDK
+
+- I am a maintainer/contributor, I want to be easily ensure that the changes made will not brake the steps required to build the projects, so that we can reduce the chance to introduce critical bugs. 
```
```suggestion
- I am a maintainer/contributor, I want to easily ensure that the changes made will not break the steps required to build the projects, so that we can reduce the chance to introduce critical bugs. 
```

asmacdo commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+#### Building samples on SDK
+
+- I am a maintainer/contributor, I want to be easily ensure that the changes made will not brake the steps required to build the projects, so that we can reduce the chance to introduce critical bugs. 
+- I am a maintainer/contributor, I want to check the result of the changes made in the scaffold by verifying the samples changes , so that I do not need to pull the pull request locally and run the steps to review the code. 
```
```suggestion
- I am a maintainer/contributor, I want to check the result of the changes made in the scaffolding by verifying the samples changes , so that I do not need to pull the pull request locally and run the steps to review the code. 
```

asmacdo commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+- I am a maintainer/contributor, I want to have the samples update with the code changes, so that I do not update the samples project after the release
+- I am a maintainer/contributor, I want to have the samples update with the code changes, so that I do not update the samples project after the release
+- I am an operator developer, I want check the examples of projects and features provided, so that I can easily know how to build my operator projects
+- I am an operator developer, I want to check the final result of the tutorial steps, so that I can easily check if what I miss when the things do not go well as expected
```
```suggestion
- I am an operator developer, I want to check the final result of the tutorial steps, so that I can easily compare if when the process does not go as expected
```

camilamacedo86 commented on enhancements/samples-generation.md:
```diff
@@ -0,0 +1,613 @@
+
+## Open Questions
+
+1. Should not the project has a Memcached GO Sample with webhooks and other without using webhooks since it shows an advanced feature and it used is not mandatory? Note the Go tutorial do not use webhooks.  
```
Yes, we would have 2 samples. However, in upstream/kb we have none similar testdata. The idea is we have samples with the steps/implementation suggested to users in the tutorials. The Go Memcached tutorial has none webhooks and we ought to have an advanced tutorial with the additional steps to use webhooks. So, in the end, it would be:

// The Memcached sample without webhooks 
// The Memcached sample with webhooks

Note that we can re-use the first sample steps as we will do in the e2e test to build the second one. 
Also, both should be integrated with OLM since it is a key feature of the tool.

camilamacedo86 commented:
Hi @varshaprasad96 and @asmacdo,

Really thank you for your review. The typo suggestions are all accepted and the questions raised answered.  
Please, let me know if you have anything else that could be addressed here for we get it merged.

jmrodri reviewed (approved):
/lgtm
